File: src/main/java/chapter2/mergeSort/bottomupmergesort.py
def bottomUpMergeSort(anArray):
    aux = anArray.copy()
    size = 1
    while size < len(anArray):
        lo = 0
        while lo < len(anArray) - size:
            hi = min(lo + 2*size - 1, len(anArray) - 1)
            med = lo + size -1
            merge(lo, med, hi, anArray, aux)
            lo += 2 * size
        size = 2 * size


def merge(lo, med, hi, array, copy):
    idxFirst = lo
    idxSecond = med + 1
    current = lo

    for i in range(lo, hi + 1):
        copy[i] = array[i]

    while current <= hi:
        if idxFirst > med:
            array[current] = copy[idxSecond]
            idxSecond +=1
        elif idxSecond > hi:
            array[current] = copy[idxFirst]
            idxFirst +=1
        elif copy[idxFirst] < copy[idxSecond]:
            array[current] = copy[idxFirst]
            idxFirst +=1
        else:
            array[current] = copy[idxSecond]
            idxSecond +=1
        current+=1
        

File: src/main/java/chapter2/mergeSort/BottomUpMergeSort.java
package chapter2.mergeSort;

public class BottomUpMergeSort {
    public static void main(String[] args) {
		Integer[] array = new Integer[]{0, -2, -5, 3, 18, 9, -9, -43, 12, 1993, 37, 28, 1};
		sort(array);
		System.out.println(print(array));
    }

    public static void sort(Comparable[] array) {
		Comparable[] aux = new Comparable[array.length];
		for (int size = 1; size < array.length; size = 2 * size) { // Increment size in power of 2
			for (int lo = 0; lo < array.length - size; lo += size + size) { // In each iteration, we have 2 arrays of size 2; Increment lo in 2 sizes
			merge(array, aux, lo, lo + size - 1, Math.min(lo + size + size -1, array.length - 1));
			}
		}
    }

    public static void merge(Comparable[] array, Comparable[] aux, int low, int mid, int high) {
		int idxFirstPart = low;
		int idxSecondPart = mid + 1;

		for (int k = low; k <= high; k++) aux[k] = array[k];

		for (int j = low; j <= high; j++) {
			if (idxFirstPart == mid + 1) {
			array[j] = aux[idxSecondPart++];
			} else if (idxSecondPart == high + 1) {
			array[j] = aux[idxFirstPart++];
			} else if (less(aux[idxFirstPart], aux[idxSecondPart])) {
			array[j] = aux[idxFirstPart++];
			} else  {
			array[j] = aux[idxSecondPart++];
			}
		}
    }

    public static boolean less(Comparable a, Comparable b) {
		return a.compareTo(b) < 0;
    }
    
    public static String print(Comparable[] array) {
		StringBuilder str = new StringBuilder();
		for (Comparable value : array) {
			str.append(value);
			str.append(" ");
		}
		return str.toString();
    }
}
